### PR TITLE
*: fix the problem that PointGet returns wrong results in the case of overflow

### PR DESCRIPTION
**Pull request by zimulala**
### What problem does this PR solve? 
```
tidb> create table t6 (a bigint);
Query OK, 0 rows affected (0.00 sec)
tidb> insert into t6 values(9223372036854775807);
Query OK, 1 row affected (0.00 sec)
tidb> alter table t6 add unique key idx(a);
Query OK, 0 rows affected (0.04 sec)
```
Before this PR:
```
tidb> select * from t6 where a = 9223372036854775808;
+---------------------+
| a                   |
+---------------------+
| 9223372036854775807 |
+---------------------+
1 row in set, 1 warning (0.01 sec)
```
After this PR:
```
tidb>  select * from t6 where a = 9223372036854775808;
Empty set (0.00 sec)
```

### What is changed and how it works?
Do `ConvertTo` to the point-get value. Handling the overflow and truncated errors.

### Check List 

Tests 

 - Unit test

Related changes

 - Need to cherry-pick to the release branch

Release note

 - Write release note for bug-fix or new feature.

### CODE REVIEW

**Comment by zimulala** [comment]
PTAL @lzmhhh123

**Comment by zimulala** [comment]
PTAL @lzmhhh123 @XuHuaiyu

File: planner/core/point_get_plan.go [hunk 1]
```diff
@@ -652,6 +653,9 @@ func tryPointGetPlan(ctx sessionctx.Context, selStmt *ast.SelectStmt) *PointGetP
 		return nil
 	}
 	pi := tbl.GetPartitionInfo()
+	if pi != nil && pi.Type != model.PartitionTypeHash {
```
[the remark this answers is not in the document]
Comment (zimulala) [reply planner/core/point_get_plan.go:656]: Yes, it's the original code, I only move it.

File: planner/core/point_get_plan.go [hunk 2]
```diff
@@ -901,17 +891,45 @@ func getNameValuePairs(nvPairs []nameValuePair, tblName model.CIStr, expr ast.Ex
...
+			if !terror.ErrorEqual(types.ErrTruncatedWrongVal, err) {
+				return nil, false
+			}
+		}
+		// The converted result must be same as original datum.
+		cmp, err := d.CompareDatum(stmtCtx, &dVal)
+		if err != nil {
+			return nil, false
+		} else if cmp != 0 {
+			// return nil, true
```
Comment (crazycs520) [inline planner/core/point_get_plan.go:926]: ?

File: planner/core/point_get_plan.go (hunk 1 above)
[the remark this answers is not in the document]
Comment (zimulala) [reply planner/core/point_get_plan.go:656]: Now, we only support for `HashPartition`.

File: planner/core/point_get_plan.go (hunk 2 above)
Comment (zimulala) [reply planner/core/point_get_plan.go]: @crazycs520 Done.

**Comment by zimulala** [comment]
PTAL @crazycs520 @lzmhhh123 @XuHuaiyu

File: planner/core/point_get_plan.go [hunk 3]
```diff
@@ -901,17 +891,44 @@ func getNameValuePairs(nvPairs []nameValuePair, tblName model.CIStr, expr ast.Ex
...
+		if tbl.IsView() {
+			return nil, false
 		}
 		if colName.Name.Table.L != "" && colName.Name.Table.L != tblName.L {
-			return nil
+			return nil, false
+		}
+		col := model.FindColumnInfo(tbl.Cols(), colName.Name.Name.L)
+		if col == nil || // Handling the case when the column is _tidb_rowid.
+			(col.Tp == mysql.TypeString && col.Collate == charset.CollationBin) { // This type we needn't to pad `\0` in here.
```
Comment (crazycs520) [inline planner/core/point_get_plan.go:908]: Any test case for that? I don't know this is used for what.

Comment (zimulala) [reply planner/core/point_get_plan.go:908]: Yes. These tests already exist. The test case for `create table t(a binary(2) primary key, b binary(2));` table, like ` select * from t where a = "a  ";`
@crazycs520

**Comment by zimulala** [comment]
PTAL 
@lzmhhh123 @XuHuaiyu  @crazycs520 @bb7133

**Comment by wjhuang2016** [comment]
mysql> select * from t6 where a != 9223372036854775808;
ERROR 1690 (22003): constant 9223372036854775808 overflows bigint

PTAL @zimulala

**Comment by zimulala** [comment]
@wjhuang2016 
This PR fixed the problem that PointGet returns wrong results in the case of overflow. The specific scene of the repair is stated in the description. The scene you mentioned has not been changed before or after this PR. You can create a new issue.

**Pull request merged**